### ISSUE TITLE
`for..in` object type keys

### ISSUE THREAD

**Issue by blaumeise20** [microsoft/TypeScript#48745]
# Bug Report

### 🔎 Search Terms

for..in object

### 🕗 Version & Regression Information

- I think it always was like that

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBCBGArGBeGBvAUDHMDWApgJ4CMAXDAEQAWBANnSJQDTa6FEBMFJnALKwC+AbkyYAZiABOMABShIsDjACWYOEgCUGNjgUQQdAgDpGAc1kJEAbQ4BdTaMFA)

### 💻 Code

```ts
const obj = {
    key1: "hello",
    key2: 124,
};

for (const key in obj) {
    console.log(obj[key]);
}
```

### 🙁 Actual behavior

There is an error "Element is implicitly 'any' because ... cannot be indexed with type 'string'".
`key` is of type string.

### 🙂 Expected behavior

It should work, and `key` should be `"key1" | "key2"`.

**Comment by MartinJohns** [comment]
Objects are not sealed. Your expectation would be unsound.

**Comment by blaumeise20** [comment]
Hm I don't get what you mean. You can't add new properties like `obj.abc = true`, can you?

If that is possible, why not use `keyof Whatever` as the type? Because I can guarantee that `obj` will have a key as the value in the `key` variable.

**Comment by MartinJohns** [comment]
See #44706.

**Comment by fatcerberus** [comment]
Just for posterity:
```
interface FooPlus { foo: number, bar: string, qux: string }
interface Foo { foo: number, bar: string };

const foobar: FooPlus = { foo: 42, bar: "foo", qux: "moo" };
const foo: Foo = foobar;

for (const key in foo) {
    console.log(obj[key]);
}
```
`key` shouldn't be type `"foo" | "bar"` here because it will also take the value `"qux"`.

**Comment by blaumeise20** [comment]
Oh, the problem is the edge case with more keys! I understand. But, is something like this possible without casts or ts-ignore? Because this is perfectly valid JS:
```ts
const defaultConfig = { someprop: true, abc: 123 };
const currentConfig = { someprop: true, abc: 456 };

const newConfig = Object.assign({}, currentConfig);
for (const key in newConfig) {
    if (newConfig[key] === defaultConfig[key])
        delete newConfig[key];
}
```

**Comment by MartinJohns** [comment]
Having more properties is hardly an edge case. Using a type assertion is an appropriate solution here.

**Comment by fatcerberus** [comment]
My experience has been, iterating over keys is generally paired with a string index signature/`Record<string, T>`, in which case the typing of the key as `string` isn’t an issue.  If you know the shape of your object in advance enough to name the keys in the source upfront, you don’t typically need to iterate over it programmatically.

**Comment by blaumeise20** [comment]
That is true, but I want to be able to dynamically add properties to `defaultConfig`. I also created a type `type Config = typeof defaultConfig`, so I just have to add a new property to the default, and I'm done.

You think I should write this?
```ts
if (newConfig.someprop == defaultConfig.someprop) delete newConfig.someprop;
if (newConfig.abc == defaultConfig.abc) delete newConfig.abc;
```
